### ISSUE TITLE
Dart ignores errors in SSL_read

### ISSUE THREAD

**Issue by brianquinlan** [dart-lang/sdk#48311]
The relevant code is here:
https://github.com/dart-lang/sdk/blob/250173e60cbefc56a3130da5ad6187ae61f8a34c/runtime/bin/secure_socket_filter.cc#L723

Notice how the errors in `SSL_read` are (possibly) logged but no other action is taken.

**Comment by brianquinlan** [comment]
This is a repro:

```dart
import 'dart:io';
import 'dart:convert';

Future<void> main() async {
//  await useHttpClient();
  print('>>>>>>>>>>>>>>>>>>>>>');
  final secureSocket = await SecureSocket.connect("test-e-kasy.mf.gov.pl", 443,
      onBadCertificate: (certificate) => true,
      supportedProtocols: ["http/1.1"]);

  secureSocket.setOption(SocketOption.tcpNoDelay, true);
  print('remoteAddress: ${secureSocket.remoteAddress}');
  print('remotePort: ${secureSocket.remotePort}');
  print('peerCertificate.subject: ${secureSocket.peerCertificate!.subject}');
  print('protocol: ${secureSocket.selectedProtocol}');

  final s = secureSocket.listen((data) {
    print(utf8.decode(data));
  });

  print('Sending request...');
  secureSocket.write("GET /api/Repository/Ping HTTP/1.1\r\n");
  secureSocket.write("Host: test-e-kasy.mf.gov.pl\r\n");
  secureSocket.write("User-Agent: curl/7.64.1\r\n");
  secureSocket.write("Accept: */*\r\n\r\n");
  print('Listening...');

  try {
    await s.asFuture();
    } on TlsException catch (e) {
      print('This is an exception: $e [${e.message} ${e.osError}]');
    }
}
```

That will fail with:
```
SocketException: OS Error: Connection reset by peer, errno = 54, address = test-e-kasy.mf.gov.pl, port = 62965
```

The actual failure should be `TlsException`.

**Comment by brianquinlan** [comment]
Fixed in 98634512f240a7e4db89f44df7dd488b684a548f
